8290529: C2: assert(BoolTest(btest).is_canonical()) failure

rwestrel opened the pull request:
For the test case:

1) In Parse::do_if(), tst0 is:

(Bool#lt (CmpU 0 Parm0))

2) transformed by gvn in tst:

(Bool#gt (CmpU Parm0 0))

3) That test is not canonical and is negated and retransformed which
results in:

(Bool#eq (CmpI Parm0 0))

The assert fires because that test is not canonical either.

The root cause I think is that the (CmpU .. 0) -> (CmpI .. 0) only
triggers if the condition of the CmpU is canonical (and results in a
non canonical test). Tweaking it so it applies even if the condition
is not leads to the following change in the steps above:

2) (Bool#ne (CmpI Parm0 0))

which is a canonical test.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290529](https://bugs.openjdk.org/browse/JDK-8290529): C2: assert(BoolTest(btest).is_canonical()) failure


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [8850b36c](https://git.openjdk.org/jdk/pull/9553/files/8850b36c9241ffb4c897fc79810f18c01169a19c)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [8850b36c](https://git.openjdk.org/jdk/pull/9553/files/8850b36c9241ffb4c897fc79810f18c01169a19c)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9553/head:pull/9553` \
`$ git checkout pull/9553`

Update a local copy of the PR: \
`$ git checkout pull/9553` \
`$ git pull https://git.openjdk.org/jdk pull/9553/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9553`

View PR using the GUI difftool: \
`$ git pr show -t 9553`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9553.diff">https://git.openjdk.org/jdk/pull/9553.diff</a>

</details>

vnkozlov commented on src/hotspot/share/opto/subnode.cpp:
```diff
@@ -1621,12 +1621,12 @@ Node *BoolNode::Ideal(PhaseGVN *phase, bool can_reshape) {
   // Change x u< 1 or x u<= 0 to x == 0
```
Update comment.

vnkozlov replied:
What about updating this comment.

rwestrel replied:
Right. Sorry I forgot about that one. I updated the comment.

vnkozlov reviewed (commented):
Changes are good. If possible add IR framework test.

rwestrel commented:
> Changes are good. If possible add IR framework test.

Thanks for looking a this. Are you asking that the test that I added to catch this specific problem be turned into and IR framework test or are you asking for a separate test to verify the transformation in general?

vnkozlov commented:
> > Changes are good. If possible add IR framework test.
> 
> Thanks for looking a this. Are you asking that the test that I added to catch this specific problem be turned into and IR framework test or are you asking for a separate test to verify the transformation in general?

Separate IR test to verify the transformation in general.

rwestrel commented:
> > > Changes are good. If possible add IR framework test.
> > 
> > 
> > Thanks for looking a this. Are you asking that the test that I added to catch this specific problem be turned into and IR framework test or are you asking for a separate test to verify the transformation in general?
> 
> Separate IR test to verify the transformation in general.

I added one.

vnkozlov reviewed (approved):
Good.

TobiHartmann reviewed (approved):
Looks good to me.

So the root cause is JDK-8276162, correct?

rwestrel commented:
> So the root cause is JDK-8276162, correct?

It's the one that is causing it to show up now but JDK-8209544 is the one that introduced the bug.

rwestrel commented:
@vnkozlov @TobiHartmann thanks for the reviews

rwestrel commented:
For the record: Test CmpUWithZero.java requires the compareUnsigned intrinsic which in turn requires the CmpU3 only implemented on x86_64 so I made that test x86_64 only.

TobiHartmann commented:
Thanks for the clarifications.

rwestrel commented:
/integrate
